xds/csds/csds_test.go

```diff
@@ -0,0 +1,672 @@
+	endpoints    = make([]*v3endpointpb.ClusterLoadAssignment, len(edsTargets))
+	endpointAnys = make([]*anypb.Any, len(edsTargets))
+	ips          = []string{"0.0.0.0", "1.1.1.1"}
+	ports        = []uint32{123, 456}
+)
+
+func init() {
+	for i := range ldsTargets {
+		listeners[i] = e2e.DefaultListener(ldsTargets[i], rdsTargets[i])
+		listenerAnys[i], _ = ptypes.MarshalAny(listeners[i])
+	}
+	for i := range rdsTargets {
+		routes[i] = e2e.DefaultRouteConfig(rdsTargets[i], ldsTargets[i], cdsTargets[i])
+		routeAnys[i], _ = ptypes.MarshalAny(routes[i])
+	}
+	for i := range cdsTargets {
+		clusters[i] = e2e.DefaultCluster(cdsTargets[i], edsTargets[i])
+		clusterAnys[i], _ = ptypes.MarshalAny(clusters[i])
+	}
+	for i := range edsTargets {
+		endpoints[i] = e2e.DefaultEndpoint(edsTargets[i], ips[i], ports[i])
+		endpointAnys[i], _ = ptypes.MarshalAny(endpoints[i])
+	}
+}
+
+func TestCSDS(t *testing.T) {
+	const retryCount = 10
+
+	xdsC, mgmServer, nodeID, stream, cleanup := commonSetup(t)
+	defer cleanup()
+
+	for _, target := range ldsTargets {
+		xdsC.WatchListener(target, func(client.ListenerUpdate, error) {})
+	}
+	for _, target := range rdsTargets {
+		xdsC.WatchRouteConfig(target, func(client.RouteConfigUpdate, error) {})
+	}
+	for _, target := range cdsTargets {
+		xdsC.WatchCluster(target, func(client.ClusterUpdate, error) {})
+	}
+	for _, target := range edsTargets {
+		xdsC.WatchEndpoints(target, func(client.EndpointsUpdate, error) {})
+	}
+
+	for i := 0; i < retryCount; i++ {
+		err := checkForRequested(stream)
+		if err == nil {
+			break
+		}
+		if i == retryCount-1 {
+			t.Fatalf("%v", err)
+		}
+		time.Sleep(time.Millisecond * 100)
+	}
+
+	if err := mgmServer.Update(e2e.UpdateOptions{
+		NodeID:    nodeID,
+		Listeners: listeners,
+		Routes:    routes,
+		Clusters:  clusters,
+		Endpoints: endpoints,
+	}); err != nil {
+		t.Fatal(err)
+	}
+	for i := 0; i < retryCount; i++ {
+		err := checkForACKed(stream)
+		if err == nil {
+			break
+		}
+		if i == retryCount-1 {
+			t.Fatalf("%v", err)
+		}
+		time.Sleep(time.Millisecond * 100)
+	}
+
+	const nackResourceIdx = 0
+	if err := mgmServer.Update(e2e.UpdateOptions{
+		NodeID: nodeID,
+		Listeners: []*v3listenerpb.Listener{
+			{Name: ldsTargets[nackResourceIdx], ApiListener: &v3listenerpb.ApiListener{}}, // 0 will be nacked. 1 will stay the same.
+		},
+		Routes: []*v3routepb.RouteConfiguration{
+			{Name: rdsTargets[nackResourceIdx], VirtualHosts: []*v3routepb.VirtualHost{{
+				Routes: []*v3routepb.Route{{}},
+			}}},
+		},
+		Clusters: []*v3clusterpb.Cluster{
+			{Name: cdsTargets[nackResourceIdx], ClusterDiscoveryType: &v3clusterpb.Cluster_Type{Type: v3clusterpb.Cluster_STATIC}},
+		},
+		Endpoints: []*v3endpointpb.ClusterLoadAssignment{
+			{ClusterName: edsTargets[nackResourceIdx], Endpoints: []*v3endpointpb.LocalityLbEndpoints{{}}},
+		},
+		SkipValidation: true,
+	}); err != nil {
+		t.Fatal(err)
+	}
+	for i := 0; i < retryCount; i++ {
+		err := checkForNACKed(nackResourceIdx, stream)
```

Comment:
Same here? We should receive a callback error once the resource is NACKed, right? We can again block on that channel and then call `checkForNACKed` just once.